models/attachment.go

```diff
@@ -1,11 +1,142 @@
+	vc := ValidationContext{
+		FromAddress: "foo@bar.com",
+		BaseURL:     "http://example.com",
+	}
+	td := Result{
+		BaseRecipient: BaseRecipient{
+			Email:     "foo@bar.com",
+			FirstName: "Foo",
+			LastName:  "Bar",
+			Position:  "Test",
+		},
+		RId: "123456",
+	}
+	ptx, err := NewPhishingTemplateContext(vc, td.BaseRecipient, td.RId)
+	if err != nil {
+		return err
+	}
+	_, err = a.ApplyTemplate(ptx)
+	return err
+}
+
+// ApplyTemplate parses different attachment files and applies the supplied phishing template.
+func (a *Attachment) ApplyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {
+
+	decodedAttachment := base64.NewDecoder(base64.StdEncoding, strings.NewReader(a.Content))
+
+	// If we've already determined there are no template variables in this attachment return it immediately
+	if a.vanillaFile == true {
+		return decodedAttachment, nil
+	}
+
+	// Decided to use the file extension rather than the content type, as there seems to be quite
+	//  a bit of variability with types. e.g sometimes a Word docx file would have:
+	//   "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
+	fileExtension := filepath.Ext(a.Name)
+
+	switch fileExtension {
+
+	case ".docx", ".docm", ".pptx", ".xlsx", ".xlsm":
+		// Most modern office formats are xml based and can be unarchived.
+		// .docm and .xlsm files are comprised of xml, and a binary blob for the macro code
+
+		// Zip archives require random access for reading, so it's hard to stream bytes. Solution seems to be to use a buffer.
+		// See https://stackoverflow.com/questions/16946978/how-to-unzip-io-readcloser
+		b := new(bytes.Buffer)
+		b.ReadFrom(decodedAttachment)
+		zipReader, err := zip.NewReader(bytes.NewReader(b.Bytes()), int64(b.Len())) // Create a new zip reader from the file
+
+		if err != nil {
+			return nil, err
+		}
+
+		newZipArchive := new(bytes.Buffer)
+		zipWriter := zip.NewWriter(newZipArchive) // For writing the new archive
+
+		// i. Read each file from the Word document archive
+		// ii. Apply the template to it
+		// iii. Add the templated content to a new zip Word archive
+		for _, zipFile := range zipReader.File {
+			ff, err := zipFile.Open()
+			if err != nil {
+				return nil, err
+			}
+			defer ff.Close()
+			contents, err := ioutil.ReadAll(ff)
+			if err != nil {
+				return nil, err
+			}
+			subFileExtension := filepath.Ext(zipFile.Name)
+			var tFile string
+			if subFileExtension == ".xml" || subFileExtension == ".rels" { // Ignore other files, e.g binary ones and images
+				// For each file apply the template.
+				tFile, err = ExecuteTemplate(string(contents), ptx)
+				if err != nil {
+					zipWriter.Close() // Don't use defer when writing files https://www.joeshaw.org/dont-defer-close-on-writable-files/
+					return nil, err
+				}
+				// Check if the subfile changed. We only need this to be set once to know in the future to check the 'parent' file
+				if tFile != string(contents) {
+					a.vanillaFile = true
```

Comment:
So I have some other changes that I want to make in order to make this a bit simpler, but to start I wanted to get your take on this.

I added a line in the test that says:

```go
c.Assert(a.vanillaFile, check.Equals, strings.Contains(fname, "without-vars"))
```

And it seems to have failed on `word-file-with-vars.docm.b64`:

```
attachment_test.go:66:
    c.Assert(a.vanillaFile, check.Equals, strings.Contains(fname, "without-vars"))
... obtained bool = true
... expected bool = false
```

That said, I think this should be `a.vanillaFile = false`? But then that seems to break on `word-file-without-vars.docx.b64`. All this to say I think we should make sure we have adequate testing. Preferably, we'd have tests to both validate that templates are being applied, being applied correctly, and that processing a templated attachment twice indeed gives the right output.

Reply:
I notice that rolling back to the previous commit results in the tests passing:

Test cases:
```
c.Assert(a.vanillaFile, check.Equals, strings.Contains(fname, "without-vars"))
c.Assert(a.vanillaFile, check.Not(check.Equals), strings.Contains(fname, "with-vars"))
```

Revert:
```
git checkout 304aa3b6b7afd4e58ce8c90fced19776318cd578  #Attempt at keeping objects as Readers
go test -v -timeout 30s github.com/gophish/gophish/models

---- SNIP ----
Checking attachment file -> html-file-with-vars.html
Checking attachment file -> html-file-without-vars.html
Checking attachment file -> text-file-with-vars.txt
Checking attachment file -> text-file-without-vars.txt
Checking attachment file -> word-file-with-vars.docm.b64
Checking attachment file -> word-file-without-vars.docx.b64
---- SNIP ----

ok  	github.com/gophish/gophish/models	(cached)
```

So, something we've tweaked in the last commit 'Small fixes' has broken something.

You're correct, the `isVanilla` value for `word-file-with-vars.docm.b64` should be `false` - it seems the Word files have become inverted somehow (they are correctly reflected in the previous commit).

I'll do some digging and see if I can find the cause, and also give some thought to the additional tests you suggest.

Reply:
Aha, found it! I misread your point: 
>  That said, I think this should be `a.vanillaFile = false` ?

Yes you're correct, the assignment should be `a.vanillaFile = false` and the missing piece was that before we process the file we set `a.vanillaFile` to `true` (assume it's vanilla unless the processing proves otherwise).

We now pass both tests mentioned above:

```
go test -timeout 30s github.com/gophish/gophish/models

ok  	github.com/gophish/gophish/models	0.320s
```

Reply:
> All this to say I think we should make sure we have adequate testing. Preferably, we'd have tests to both validate that templates are being applied, being applied correctly, and that processing a templated attachment twice indeed gives the right output.

I'm working on this now. My plan is to create 'expected' output files for each file, so we can compare input and output. Finding some initial oddities with `.docx` files - with no templates, the before and after files look identical when decompressed, but their zip sizes are off by almost double. I think it's an artefact of different compression routines. I'll figure it out and report back.

Reply:
I created some 'expected' output files, that have been parsed by the templating mechanism and give the desired outcome (I've manually expected them):

```
html-file-with-vars.templated.html
html-file-without-vars.templated.html
text-file-with-vars.templated.txt
text-file-without-vars.templated.txt
word-file-with-vars.docm.templated.b64
word-file-without-vars.docx.templated.b64
```

The plain text files are obvious. The .b64 ones can be decoded and inspected like so:

```
base64 --decode word-file-with-vars.docm.templated.b64 > word-file-with-vars.docm 
```

<img width="640" alt="Screenshot 2020-12-10 at 18 27 36" src="https://user-images.githubusercontent.com/3966613/101807563-86fcd280-3b15-11eb-904a-99ed39c66c7c.png">

We then compare the output of applying the template to a file, with the expected outcome file:

```
c.Assert(templatedFile, check.Equals, expectedOutput)
```

Seems to run fine:

```
go test -timeout 30s github.com/gophish/gophish/models
ok  	github.com/gophish/gophish/models	0.343s
```

Next I'll have a go at creating a test file for applying the template twice to a file.

### Also, something else to bring up:

I noticed when creating these test files that when we unarchive a Word file, process it, and re-archive it the resultant file is much smaller. The reason seems to be that Microsoft only applies about 11% compression, whereas the go zip library will apply about 65% (for these test files at least). I don't think this is a big issue for us, but we might want to look into lowering the compression ratio to better match Microsoft's (MS does this to have faster opening/saving I assume). Also, we might want to return the original file rather than the uncompressed and recompressed one if no template variables are present. If we don't, the first email will have slightly smaller Word document. Not a biggie, but maybe better to be consistent and deliver the exact file the user attached.

I could achieve that by adding the following to the [end of the word processing code](https://github.com/gophish/gophish/pull/1936/files#diff-1ed3f192802af9040844eed631caaa322821342e401dc8d4804bc2ec11e0ae8aR123)
```
if a.vanillaFile == true {
    return base64.NewDecoder(base64.StdEncoding, strings.NewReader(a.Content)), nil // If no template variables were found, return the original file
}
```

Reply:
> processing a templated attachment twice indeed gives the right output

Hi @jordan-wright just checking what you mean by this. Either:

(a) Apply templating function twice to the same input, to make sure we get the same output both times; or
(b) Apply templating function twice in sequence, so, `ApplyTemplate(ApplyTemplate(input))` ?

I don't think we need (a) as we now test the generated output against expected output. For (b) I that's functionally equivalent to the existing test cases with no variables in them (but I could add this test if you like).

I've also added a PowerPoint and Excel test case.